go/database/mpt/verification.go

```diff
@@ -570,6 +622,72 @@ func verifyHashesStoredWithParents[N any](
 	return nil
 }
 
+// verifyContractCodes runs list of validation checks on Mpt and on the
+// code file stored in the given directory. These checks include:
+// 1) Fatal checks
+// - All CodeHashes within the Mpt are present in the code file
+// - All CodeHashes within the code file are correct
+// 2) Non-fatal checks
+// - There are no extra Code Hashes not referenced by any account
+func verifyContractCodes(directory string, source *verificationNodeSource, observer VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Checking contract codes ..."))
+
+	codeFile := filepath.Join(directory, "codes.dat")
+	codes, err := readCodes(codeFile)
+	if err != nil {
+		return err
+	}
+
+	found := make(map[common.Hash]bool)
+	check := func(acc *AccountNode) error {
+		codeHash := acc.info.CodeHash
+		// skip accounts that are not contracts
+		if codeHash == emptyCodeHash {
+			return nil
+		}
+		// no need to check the hash correctness more than once
+		if _, exists := found[codeHash]; exists {
+			return nil
+		}
+
+		// check that the code hash is present in the code file
+		byteCode, exists := codes[codeHash]
+		if !exists {
+			return fmt.Errorf("the hash %x is present in the mpt but missing from the code file", codeHash)
+		}
+		// check correctness of the code hash
+		if got, want := common.Keccak256(byteCode), &codeHash; got.Compare(want) != 0 {
+			return fmt.Errorf("unexpected code hash for address, got: %x want: %x", got, want)
+		}
+		// mark a found hash
+		found[codeHash] = true
```

Comment:
I think this should go before the `if` condition - if there is a corrupted contract used in many accounts, it would print it many times now

Reply:
Since it return an error, the verification is fatal hence it ends no?

Reply:
I think it iterates over all accounts first, and only then accumulate errors, here:

https://github.com/Fantom-foundation/Carmen/blob/petr/verify-codes/go/database/mpt/verification.go#L900-L902

Reply:
You are right. Good catch.